### PR TITLE
Fix kl_div returning nan instead of inf at (inf, 0)

### LINKED ISSUE #38239: jax.scipy.special.kl_div returns nan instead of inf at x=inf, y=0.0
### Description

`jax.scipy.special.kl_div(inf, 0.0)` returns `nan` instead of `inf`.

The KL divergence is defined as:

```
kl_div(x, y) = x*log(x/y) - x + y   if x > 0 and y > 0
             = y                      if x == 0 and y >= 0
             = inf                    otherwise
```

With `x=inf` and `y=0`, neither of the first two conditions holds, so the
correct result is `inf`. SciPy returns `inf` correctly. JAX returns `nan`.

## Origin

Found via testing derived from:

```
jax/tests/lax_scipy_special_functions_test.py
  ::LaxScipySpecialFunctionsTest::testScipySpecialFun_kl_div
```

The existing test uses `rand_positive` inputs only, which never include
`x=inf` or `y=0`. The property uses a **reference oracle**:
`jax.scipy.special.kl_div(x, y)` must agree with
`scipy.special.kl_div(x, y)` in NaN/inf classification for all inputs.

Related: #27088 (state of `jax.scipy.special` functions).

## Verification (SciPy is correct)

The mathematical definition gives `+inf` for `x=inf, y=0`. The dominant
term `x*log(x/y)` diverges, confirmed at finite approximations:

```python
import numpy as np
for x in [1e3, 1e6, 1e9, 1e12]:
    y = 1e-10
    print(f"kl_div({x:.0e}, {y:.0e}) = {x * np.log(x/y) - x + y:.4e}")
# kl_div(1e+03, 1e-10) = 2.8934e+04
# kl_div(1e+06, 1e-10) = 3.5841e+07
# kl_div(1e+09, 1e-10) = 4.2749e+10
# kl_div(1e+12, 1e-10) = 4.9657e+13
```

The limit is `+∞`. SciPy is correct; the bug is in JAX.

## Minimal reproducer

```python
import numpy as np
import scipy.special
import jax
import jax.scipy.special

jax.config.update("jax_enable_x64", True)

x = np.float32(np.inf)
y = np.float32(0.0)

print(scipy.special.kl_div(x, y))                                    # inf  ← correct
print(jax.scipy.special.kl_div(x, y))                                # nan  ← BUG
print(jax.jit(jax.scipy.special.kl_div)(
    jax.numpy.array(x), jax.numpy.array(y)))                         # nan  ← BUG
```

## Expected behavior

```
inf
inf
inf
```

## Actual behavior

```
inf
nan
nan
```

## Additional cases

The bug is specific to `x=inf, y=0`. All other tested combinations are
consistent with SciPy:

```
x=inf,  y=0.0:  scipy=inf, jax=nan  ← BUG
x=inf,  y=0.5:  scipy=nan, jax=nan  OK
x=inf,  y=inf:  scipy=nan, jax=nan  OK
x=0.0,  y=0.0:  scipy=0.0, jax=0.0 OK
x=-1.0, y=0.5:  scipy=inf, jax=inf  OK
x=nan,  y=0.5:  scipy=nan, jax=nan  OK
```

## System info

```
jax:     0.10.1
jaxlib:  0.10.1
numpy:   2.4.6
scipy:   1.17.1
python:  3.12.3
OS:      Linux-6.17.0-29-generic-x86_64-with-glibc2.39
backend: cpu
```

### PR DESCRIPTION
**Pull request by factnn**
## Summary

Fixes #38239: `jax.scipy.special.kl_div(inf, 0.0)` returns `nan` instead of `inf`.

## Root Cause

`kl_div` delegates to `rel_entr(p, q) - p + q`. At `p=inf, q=0`, `rel_entr(inf, 0)` correctly returns `inf`, but then `inf - inf + 0 = nan`.

## Fix

Use piecewise conditions per the definition instead of raw arithmetic:
- `p > 0, q > 0` → `rel_entr(p, q) - p + q`
- `p == 0, q >= 0` → `q`
- otherwise → `inf`

This avoids computing `inf - inf` outside the valid domain.

## Changes

- `jax/_src/scipy/special.py`: 4 lines

Fixes #38239

* scipy
  * Fixed `kl_div(inf, 0.0)` returning nan instead of inf ([#38239](https://github.com/jax-ml/jax/issues/38239))

### CODE REVIEW

**Comment by factnn** [comment]
@jakevdp I'm guessing this falls into the same category as #38267 — the fix should be at the XLA level rather than adding edge-case handling at the JAX layer. Should I close this one too?

**Comment by jakevdp** [comment]
No, `kl_div` is not an XLA function. This fix should be on the JAX side.

That said, there are some fundamental issues with this PR regarding handling of NaN inputs, and those will have to be addressed. The simple fix offered by gemini above is probably not the best approach, because it adds so many extra layers of masking. Better I think would be to rely on correct NaN handling in `rel_entr`, but it looks like that will have to be fixed first.

**Comment by jakevdp** [comment]
Sorry for being so slow to review this: we've been pretty overwhelmed by the recent flood of AI-driven bug reports and fixes to special function corner cases like this one. People generate them in seconds, but it takes a lot of attention by human maintainers to decide on the right course of action and evaluate the proposed fixes.

**Comment by factnn** [comment]
@jakevdp Got it, that makes sense. So the sequence would be: first fix NaN handling in `rel_entr`, then `kl_div` can use the simpler `rel_entr(p,q) - p + q` without extra masking? Happy to work on that — could you point me at what specifically needs fixing in `rel_entr`?

**Comment by factnn** [comment]
> Sorry for being so slow to review this: we've been pretty overwhelmed by the recent flood of AI-driven bug reports and fixes to special function corner cases like this one. People generate them in seconds, but it takes a lot of attention by human maintainers to decide on the right course of action and evaluate the proposed fixes.

I can't agree more. Since I'm a maintainer of my org's repos too, and we recently had a flood of AI PRs that overwhelmed our CI and reviewers... Thanks for the guidance across all these PRs, I'll keep contributing.

**Comment by factnn** [comment]
Closing this — falls into the same category as #38267, not a meaningful fix to chase. Will focus on real bugs instead.

**Comment by jakevdp** [comment]
This falls under #38651; please note that it may take a while for this to receive attention from the team.
